### PR TITLE
i#1557 cmake LOCATION: use output dirs for remaining unix instances, CMP0026

### PR DESCRIPTION
**Pull request by hgreving2304**
Incremental re-factoring towards eliminating reading LOCATION target property. The only remaining ones are part of the Windows build.

Issue: #1557

### CODE REVIEW

File: api/samples/CMakeLists.txt [hunk 1]
```diff
@@ -162,8 +161,7 @@ function (add_sample_client name source_file_list extension_list)
 
   # Provide a hint for how to use the client
   if (NOT DynamoRIO_INTERNAL OR NOT "${CMAKE_GENERATOR}" MATCHES "Ninja")
-    # XXX i#1557: replace w/ generator expression to satisfy CMP0026
-    get_target_property(path ${name} LOCATION${location_suffix})
+    DynamoRIO_get_full_name(path_test ${name})
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt]: I think you already put a comment here: s/path_test/path/.  Please verify visually that the resulting printout is correct as there is no automated check.

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,7 +63,6 @@ if ("${CMAKE_GENERATOR}" MATCHES "Visual Studio")
   else (DEBUG)
     set(CMAKE_CONFIGURATION_TYPES "RelWithDebInfo" CACHE STRING "" FORCE)
   endif (DEBUG)
-  # we want to use the _LOCATION_<config> property
```
Comment (derekbruening) [inline CMakeLists.txt]: We still use location_suffix for other vars so I think the better change is to keep the comment but remove LOCATION: sthg like `We want to use the <VARIABLE>_<CONFIG> variants of config-dependent variables.`

Ditto below.

File: make/utils_exposed.cmake [hunk 3]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
   set(${var_out} ${is_gold} PARENT_SCOPE)
 endfunction (_DR_check_if_linker_is_gnu_gold)
 
+# Takes in a target and returns the expected target name.
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: You mean the target *path*, not the name ("target name" is just that: the name within cmake, like "dynamorio", with no necessary relation to a file).

File: make/utils_exposed.cmake [hunk 4]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+#
+# XXX i#1557: DynamoRIO cmake files used to query the LOCATION target property at
+# configure time. This property has been made obsolete, see CMP0026. The function
+# here can be used to retrieve the default target name instead. However, it only
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: s/name/path/

File: make/utils_exposed.cmake [hunk 5]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+# supports the default names, as seen at configure time.
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: s/names/target directories/

File: make/utils_exposed.cmake [hunk 6]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+function(DynamoRIO_get_full_name out target)
+  get_target_property(output_name ${target} OUTPUT_NAME)
+  get_target_property(name ${target} NAME)
+  get_target_property(suffix ${target} SUFFIX)
+  get_target_property(prefix ${target} PREFIX)
+  get_target_property(type ${target} TYPE)
+  if(NOT prefix)
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: style: missing space

File: make/utils_exposed.cmake [hunk 7]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
   set(${var_out} ${is_gold} PARENT_SCOPE)
 endfunction (_DR_check_if_linker_is_gnu_gold)
 
+# Takes in a target and returns the expected target name.
+#
+# XXX i#1557: DynamoRIO cmake files used to query the LOCATION target property at
+# configure time. This property has been made obsolete, see CMP0026. The function
+# here can be used to retrieve the default target name instead. However, it only
+# supports the default names, as seen at configure time.
+function(DynamoRIO_get_full_name out target)
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: style: missing space

File: make/utils_exposed.cmake [hunk 8]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+  get_target_property(output_name ${target} OUTPUT_NAME)
+  get_target_property(name ${target} NAME)
+  get_target_property(suffix ${target} SUFFIX)
+  get_target_property(prefix ${target} PREFIX)
+  get_target_property(type ${target} TYPE)
+  if(NOT prefix)
+    set(prefix ${CMAKE_${type}_PREFIX})
+  endif()
+  if(NOT suffix)
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: style: missing space

File: make/utils_exposed.cmake [hunk 9]
```diff
@@ -145,12 +145,45 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+    set(suffix ${CMAKE_${type}_SUFFIX})
+  endif()
+  set(output_dir "")
+  get_target_property(target_type ${target} TYPE)
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: Redundant: this is already in ${type}

File: make/utils_exposed.cmake (hunk 7 above)
Comment (derekbruening) [inline make/utils_exposed.cmake]: Please rename: `DynamoRIO_get_full_path` or `DynamoRIO_get_absolute_path`

File: make/utils.cmake [hunk 10]
```diff
@@ -54,6 +54,11 @@ function (append_property_list type target name value)
   set_property(${type} ${target} PROPERTY ${name} ${value})
 endfunction (append_property_list)
 
+function(get_full_name out target)
```
Comment (derekbruening) [inline make/utils.cmake]: s/get_full_name/get_full_path/ or s/get_full_name/get_absolute_path/

File: CMakeLists.txt [hunk 11]
```diff
@@ -63,7 +63,6 @@ if ("${CMAKE_GENERATOR}" MATCHES "Visual Studio")
   else (DEBUG)
```
Comment (derekbruening) [inline CMakeLists.txt]: nit: commit title for multi-commit should have mini-descr before colon, and after should say what this particular commit does: neither is there right now.  Should be sthg like `i#1557 cmake location: use output dirs for remaining unix instances".

File: api/samples/CMakeLists.txt (hunk 1 above)
Comment (hgreving2304) [reply api/samples/CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply CMakeLists.txt]: Ok

File: make/utils_exposed.cmake (hunk 3 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Takes in a target and returns the expected full target path incl. output name.

File: make/utils_exposed.cmake (hunk 5 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Ok

File: make/utils_exposed.cmake (hunk 7 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Fixed.

File: make/utils_exposed.cmake (hunk 8 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Fixed.

File: make/utils_exposed.cmake (hunk 9 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Fixed.

File: make/utils_exposed.cmake (hunk 7 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: DynamoRIO_get_full_output_name ok? It's really not just about the path, but about including the output name

File: CMakeLists.txt (hunk 11 above)
Comment (hgreving2304) [reply CMakeLists.txt]: Edited above.

File: make/utils.cmake (hunk 10 above)
Comment (hgreving2304) [reply make/utils.cmake]: Done

File: make/utils_exposed.cmake (hunk 7 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Done

File: make/utils_exposed.cmake (hunk 6 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Fixed.

**Comment by hgreving2304** [comment]
run arm tests

**Pull request merged**